src/syntax/abs/trans.rs

```diff
@@ -0,0 +1,140 @@
+            }
+        }
+        Expr::App(app_vec) => app_vec
+            .iter()
+            .try_fold(None::<Abstract>, |result, each_expr| -> TCM<_> {
+                let abs = trans_expr_inner(each_expr, env, global_map, local_env, local_map)?;
+                Ok(match result {
+                    // First item in vec
+                    None => Some(abs),
+                    // Second or other, reduce to Right
+                    Some(left_abs) => Some(Abstract::app(left_abs, abs)),
+                })
+            })
+            // Because it's guaranteed to be non-empty.
+            .map(std::option::Option::unwrap),
+        Expr::Pipe(pipe_vec) => {
+            let mut app_vec = pipe_vec.clone();
+            app_vec.reverse();
+            trans_expr_inner(&Expr::App(app_vec), env, global_map, local_env, local_map)
+        }
+        Expr::Meta(ident) => Ok(Abstract::Meta(ident.info.clone())),
+        Expr::Cons(ident) => Ok(Abstract::Cons(ident.info.clone())),
+        Expr::ConsType(ident) => Ok(Abstract::ConsType(ident.info.clone())),
+        Expr::Bot(ident) => Ok(Abstract::Bot(ident.info.clone())),
+        Expr::Sum(expr_vec) => {
+            let abs_vec = expr_vec.iter().try_fold(Vec::new(), |mut abs_vec, expr| {
+                abs_vec.push(trans_expr_inner(
+                    expr, env, global_map, local_env, local_map,
+                )?);
+                Ok(abs_vec)
+            })?;
+            Ok(Abstract::Sum(abs_vec))
+        }
+        Expr::Pi(params, result) => {
+            let mut pi_env = local_env.clone();
+            let mut pi_map = local_map.clone();
+            let mut pi_vec: Vec<Abstract> =
+                params.iter().try_fold(Vec::new(), |pi_vec, param| {
+                    trans_pi(env, global_map, &mut pi_env, &mut pi_map, pi_vec, param)
+                })?;
+
+            // fold from right
+            pi_vec.reverse();
+            Ok(pi_vec.iter().fold(
+                trans_expr_inner(result, env, global_map, &pi_env, &pi_map)?,
+                |pi_abs, param| Abstract::pi(param.clone(), pi_abs),
+            ))
+        }
+    }
+}
+
+fn trans_pi(
+    env: &AbstractGlobalEnv,
+    global_map: &NamedEnv_<DBI>,
+    pi_env: &mut AbstractGlobalEnv,
+    pi_map: &mut NamedEnv_<DBI>,
+    mut pi_vec: Vec<Abstract>,
+    param: &Param,
+) -> TCM<Vec<Abstract>> {
+    // todo: handle implicit parameter
+    // reference implementation:
+    // https://github.com/owo-lang/OwO/blob/316e83cf532c447b03a7d210bbc3c4fc1409c861/src/type_check/elaborate.rs#L35-L64
+    assert_eq!(param.kind, ParamKind::Explicit);
+    let param_ty = trans_expr_inner(&param.ty, env, global_map, &pi_env, &pi_map)?;
+    for name in param.names.clone() {
+        let param_name = name.info.text;
+        let param_dbi: DBI = pi_env.len();
+        pi_env.insert(param_dbi, AbstractDecl::Sign(param_ty.clone()));
+        pi_map.insert(param_name, param_dbi);
```

Comment:
I am worried about this :worried:. Do you need to remove the `pi_env` elements at `param_dbi` before inserting at that position?

Reply:
Same for the code at line 25

Reply:
`pi_env` and `pi_env` should not contains `param_dbi` as a key in each iteration, how about add an assertion here?

Reply:
Can you add that?